File: 687.longest-univalue-path/solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func longestUnivaluePath(root *TreeNode) int {
    var (
        res int
        dfs func(*TreeNode, int) int
    )
    
    dfs = func(node *TreeNode, parent int) int {
        if node != nil {
            l := dfs(node.Left, node.Val)
            r := dfs(node.Right, node.Val)
            res = max(res, l + r)
            if node.Val == parent {
                return max(l, r) + 1
            }
        }
        return 0
    }

    dfs(root, -1)
    return res
}

File: 687.longest-univalue-path/solution.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun longestUnivaluePath(root: TreeNode?): Int {
        var res = 0

        fun dfs(node: TreeNode?, parentValue: Int): Int {
            if (node == null) return 0
            val left = dfs(node.left, node.`val`)
            val right = dfs(node.right, node.`val`)
            res = maxOf(res, left + right)
            return if (node.`val` == parentValue) maxOf(left, right) + 1 else 0
        }

        dfs(root, -1)
        return res
    }

}